### ISSUE TITLE
Install and configure: PHP Coding Standards Fixer

### ISSUE THREAD

**Issue by ceckoslab** [basicrum/backoffice#49]
As we are already 2 people on this project and it's expected to have more open source contributors in the future we will need a process to keep the same coding standards the same for everybody.

Please take a look at this project: https://github.com/FriendsOfPHP/PHP-CS-Fixer 
We also may look at this file as an example configuration: https://github.com/symfony/symfony/blob/master/.php_cs.dist

At the end we should have PHP-CS-Fixer installed + documented how we will use it. We also will need to run the PHP-CS-Fixer every time before we push changes to the repository.

**Comment by korzol** [comment]
Hey @ceckoslab So here is the result of some time playing with cs-fixer

Initial file 1:
```php
<?php
include_once 'config.php';
$id=$_REQUEST['addonid'];
$addon=$config['addons'][$id];
$addon['id']=$id;
$addon['urlprefix']="{$_SERVER['PHP_SELF']}?addonid=$id&amp;url=".dirname($addon['link']).'/';
$url=(empty($_REQUEST['url'])) ? $addon['link'] : $_REQUEST['url'];
$array=['1','2','3'];
$array=[
'1',
'2',
'3'
];
if($a=$b).
{
$werwe=$c;
}
include_once $url;
// php closing tag has been omitted deliberately, to avoid unwanted blank lines being sent to the browser
function name1()
{
$someVar="1";
}

function name2($arg1,$arg2)
{
$someVar="2";
return $someVar;
}
```
Fixed file 1:
```php
<?php

include_once 'config.php';
$id = $_REQUEST['addonid'];
$addon = $config['addons'][$id];
$addon['id'] = $id;
$addon['urlprefix'] = "{$_SERVER['PHP_SELF']}?addonid=$id&amp;url=".dirname($addon['link']).'/';
$url = (empty($_REQUEST['url'])) ? $addon['link'] : $_REQUEST['url'];
$array = ['1', '2', '3'];
$array = [
    '1',
    '2',
    '3',
];
if ($a = $b) {
    $werwe = $c;
}
include_once $url;
// php closing tag has been omitted deliberately, to avoid unwanted blank lines being sent to the browser
function name1()
{
    $someVar = '1';
}

function name2($arg1, $arg2)
{
    $someVar = '2';

    return $someVar;
}
```
Initial file 2:
```php
<?
declare(strict_types=1);
namespace App\BasicRum\Beacon\Catcher\Storage\File;
class Sort
{
/**
* @param array $beacons
*/
public function sortBeacons(array &$beacons)
{
// Sort the array
usort($beacons,function($element1,$element2){
return $element1[0]-$element2[0];
});
}
}
```
Fixed file 2:
```php
<?php

declare(strict_types=1);

namespace App\BasicRum\Beacon\Catcher\Storage\File;

class Sort
{
    public function sortBeacons(array &$beacons)
    {
        // Sort the array
        usort($beacons, function ($element1, $element2) {
            return $element1[0] - $element2[0];
        });
    }
}
```

In the second file you can see it also removed phpdock comment because it found it useless. It would not be removed if there would be some text besides param, like this:
Will be removed:
```php
/**
* @param array $beacons
*/
```

Wont be removed:
```php
/**
* @param array $beacons Some description text 
*/
```

We can suppress this behavior. It is controlled by *no_superfluous_phpdoc_tags* rule

I am ready to give complete documentation how to install and deploy cs-fixes. As well as bind it to pre-commit git hook.

Let me know what you think, so we can proceed further with this task

**Comment by korzol** [comment]
@ceckoslab Awaiting your approval

**Comment by ceckoslab** [comment]
Looks good @korzol 

I am going to take over this ticket in order to fix some Framework update problems.

We are having some troubles to install the code fixer with composer because of dependencies incompatibilities: 

```
 Problem 1
    - Conclusion: don't install symfony/mailer v4.4.5
    - Conclusion: don't install symfony/mailer v4.4.4
    - Conclusion: don't install symfony/mailer v4.4.3
    - Conclusion: don't install symfony/mailer v4.4.2
    - Installation request for symfony/http-kernel (locked at v4.3.9) -> satisfiable by symfony/http-kernel[v4.3.9].
    - Conclusion: don't install symfony/mailer v4.4.1
    - Conclusion: don't install symfony/mailer v4.4.0
    - Conclusion: don't install symfony/mailer v4.3.11
    - Conclusion: don't install symfony/mailer v4.3.10
    - Conclusion: don't install symfony/mailer v4.3.9
    - Conclusion: don't install symfony/mailer v4.3.8
    - Conclusion: don't install symfony/mailer v4.3.7
    - Conclusion: don't install symfony/mailer v4.3.6
    - Conclusion: don't install symfony/mailer v4.3.5
    - Conclusion: don't install symfony/mailer v4.3.4
    - Conclusion: don't install symfony/mailer v4.3.3
    - Conclusion: don't install symfony/mailer v4.3.2
    - Conclusion: don't install symfony/mailer v4.3.1
    - symfony/mailer v4.3.0 conflicts with symfony/mime[v5.0.1].
    - symfony/mailer v4.3.0 conflicts with symfony/mime[v5.0.1].
    - symfony/mailer v4.3.0 conflicts with symfony/mime[v5.0.1].
    - Installation request for symfony/mailer ^4.1 -> satisfiable by symfony/mailer[v4.3.0, v4.3.1, v4.3.10, v4.3.11, v4.3.2, v4.3.3, v4.3.4, v4.3.5, v4.3.6, v4.3.7, v4.3.8, v4.3.9, v4.4.0, v4.4.1, v4.4.2, v4.4.3, v4.4.4, v4.4.5].
    - Installation request for symfony/mime (locked at v5.0.1) -> satisfiable by symfony/mime[v5.0.1].
```

**Comment by korzol** [comment]
Here you go. This is c-fixer rules your requested
Place this into .php_cs file in project root directory

```php
<?php

if (!file_exists(__DIR__.'/src')) {
    exit(0);
}

return PhpCsFixer\Config::create()
    ->setRules([
        '@Symfony' => true,
        '@Symfony:risky' => true,
        '@PHPUnit75Migration:risky' => true,
        'php_unit_dedicate_assert' => ['target' => '5.6'],
        'array_syntax' => ['syntax' => 'short'],
        'fopen_flags' => false,
        'protected_to_private' => false,
        'combine_nested_dirname' => true,
// my additions
//        'array_syntax' => ['syntax' => 'short'], // already defined in @Symfony
        'array_indentation' => true,
        
    ])
    ->setRiskyAllowed(true)
    ->setFinder(
        PhpCsFixer\Finder::create()
            ->in(__DIR__.'/src')
            ->append([__FILE__])
            ->notPath('#/Fixtures/#')
            ->exclude([
                // directories containing files with content that is autogenerated by `var_export`, which breaks CS in output code
                // fixture templates
                'Symfony/Bundle/FrameworkBundle/Tests/Templating/Helper/Resources/Custom',
                // resource templates
                'Symfony/Bundle/FrameworkBundle/Resources/views/Form',
                // explicit trigger_error tests
                'Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/',
            ])
            // Support for older PHPunit version
            ->notPath('Symfony/Bridge/PhpUnit/SymfonyTestsListener.php')
            ->notPath('#Symfony/Bridge/PhpUnit/.*Mock\.php#')
            ->notPath('#Symfony/Bridge/PhpUnit/.*Legacy#')
            // file content autogenerated by `var_export`
            ->notPath('Symfony/Component/Translation/Tests/fixtures/resources.php')
            // test template
            ->notPath('Symfony/Bundle/FrameworkBundle/Tests/Templating/Helper/Resources/Custom/_name_entry_label.html.php')
            // explicit trigger_error tests
            ->notPath('Symfony/Component/ErrorHandler/Tests/DebugClassLoaderTest.php')
    )
;
```

**Comment by ceckoslab** [comment]
Hello @korzol 

I will test the php_cs file once we merge https://github.com/basicrum/backoffice/pull/52

**Comment by ceckoslab** [comment]
Hello @korzol 

I tested before we merged #52 and this is the command I used for a dry run:
```
vendor/bin/php-cs-fixer fix --config=.php_cs --verbose --diff --dry-run
```

Once we merge everything we will remove .php_cs from .gitignore and we will be fine to use the style fixer. 

I remember that you mentioned that we will use a pre-commit hook and we may look here if you didn't do a research yet: https://github.com/prooph/php-cs-fixer-config

**Comment by ceckoslab** [comment]
Hello @korzol 

I merged the fixes for the framework warnings. Could you add the .php_cs file to the repository and try the dry run command?

If all is fine I will merge the PR.
